File: server/users/sso.js
const _ = require('lodash');
const qs = require('qs');
const axios = require('axios');
const faculty = require('./faculty');

const SSO_URL = 'https://account.it.chula.ac.th';
const SSO_KILL = ticket => `${SSO_URL}/resources/tickets/${ticket}`;
const SSO_LOGIN = () => `${SSO_URL}/login`;
const SSO_USER = () => `${SSO_URL}/resources/users/me`;

function login(username, password) {
  return axios
    .get(SSO_LOGIN(), {
      withCredentials: true,
      params: {
        username: username.slice(0, 8),
        password: password,
        service: 'https://account.it.chula.ac.th/html',
        serviceName: 'Chula+SSO'
      }
    })
    .then(res => {
      if (_.get(res.data, 'type') == 'error') {
        throw new Error(_.get(res.data, 'content', 'SSO login error'));
      }
      const cookie = _.get(res.headers, "['set-cookie'][0]");
      const DeeTGT = _.get(qs.parse(cookie), 'DeeTGT');
      return DeeTGT;
    });
}

function validate(ticket) {
  return axios
    .get(SSO_USER(), {
      headers: {
        'accept-encoding': 'gzip;q=0,deflate,sdch',
        'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.131 Safari/537.36',
        Cookie: `DeeTGT=${ticket}`
      }
    })
    .then(res => res.data)
    .then(raw => {
      const facultyNUM = raw.ouid.slice(-2);
      const year = raw.ouid.slice(0, 2);
      return {
        ticket: ticket,
        ouid: raw.ouid,
        // pwid: encrypt(password),
        // titleTH: null,
        // titleEN: null,
        nameTH: raw.firstnameTH,
        nameEN: raw.firstname,
        surnameTH: raw.lastnameTH,
        surnameEN: raw.lastname,
        facultyNUM: +facultyNUM,
        facultyTH: faculty[facultyNUM].nameTH,
        facultyEN: faculty[facultyNUM].nameEN,
        facultyABBR: faculty[facultyNUM].nameABBR,
        year: +year
      };
    });
}

function logout(ticket) {
  return axios.delete(SSO_KILL(ticket)).then(res => res.data);
}

module.exports = { login, validate, logout };

File: sinatra/lib/sso.rb
# frozen_string_literal: true

require 'faraday'
require 'faraday_middleware'

class SSO
  @conn =
    Faraday.new(
      url: 'https://account.it.chula.ac.th',
      headers: { 'Content-Type' => 'application/x-www-form-urlencoded' }
    ) do |conn|
      conn.request :url_encoded
      conn.response :json, content_type: /\bjson$/
    end

  class << self
    attr_reader :conn
  end

  def self.ticket(username, password)
    if (username =~ /[0-9]{10}/).nil?
      raise ArgumentError, 'username must be number string length 10'
    end

    resp =
      @conn.post('/login') do |req|
        req.body = {
          username: username[0, username[0, 2].to_i > 61 ? 10 : 8],
          password: password,
          service: 'https://account.it.chula.ac.th',
          serviceName: 'ESC server'
        }
      end
    resp.headers[:location][/ticket=(.+)/, 1]
  end

  def self.validate(ticket)
    resp =
      @conn.get('/serviceValidation') do |req|
        req.params = { ticket: ticket }
        req.headers = {
          DeeAppId:
            'dc2326fef061a32bea16242be5941c7d403f485fa52fdfc69d145e3c3be2fb05',
          DeeAppSecret:
            'b3ed7ba73d5c455d9a9ab7a03cb11829ac141b4c07d535de281d244de779a6a847e430a8806ef6fb5e9bf928414e4c444465cbf0347eb5dab9b7a0a0b532e2e5'
        }
      end
    if resp.body['type'] == 'error'
      raise Unauthorized, 'username/password is wrong'
    end

    {
      id: resp.body['ouid'],
      nameTH: resp.body['firstnameth'],
      nameEN: resp.body['firstname'],
      surnameTH: resp.body['lastnameth'],
      surnameEN: resp.body['lastname'],
      faculty: resp.body['ouid'][8, 2].to_i,
      year: resp.body['ouid'][0, 2].to_i
    }
  end

  public

  def self.login!(username, password)
    SSO.validate(SSO.ticket(username, password))
  end

  def self.login(username, password)
    login!(username, password)
  rescue StandardError
    nil
  end
end
